contributors/design-proposals/local-pv-provisioner.md

```diff
@@ -0,0 +1,156 @@
+
+### Minion Provisioner
+The minion provisioner retains most key components from the previous design, including the Cache, Discoverer, and Deleter. The Populator is moved to master because it interacts with the API server. In order to interact with the master, it exposes the following gRPC calls:
+
+| RPC                      | Description |
+| --- | --- |
+| `Add/Update/RemovePV(PersistentVolume)` | Updates the Cache with the given PV. Called by master when PV informer is triggered. |
+| `PushCurrentPVs([]PersistentVolume)`    | Merges given list of PVs with the Cache. Used by the Populator. |
+| `GetPVRequest()`                        | Peeks at the next unhandled request in the queue. |
+| `CompletePVRequest(PVRequest)`          | Signals that the last request is completed and can be removed from the queue. |
+
+In the case of master failure while a request is being processed, the minion needs to keep record of the request so that a new master can process it again. Thus it's necessary for master to signal request completion.
+
+### Master Provisioner
+The populator is initialized so that it calls the appropriate gRPC for every PV Informer update. Then it populates each minion's cache by calling `PushCurrentPVs()`, passing the collection of PVs to be controlled by this minion.
+
+The master then performs minion discovery as described in the [Minion Discovery](#minion-discovery) section of design 1, and spawns a goroutine for each minion. Each goroutine periodically calls `GetPVRequest()`, performing that request if it receives one. Then it calls `CompletePVRequest()` to signal the request completion.
+
+## Design 3: Event-Driven with Custom Resources
+
+### High Level Overview
+The previous two designs are complicated by the asynchronous nature of delete and the presence of failures. 
+As recommended by @dhirajh, another design is to persist ProvisionerStatus as a Custom Resource, and to drive all operations with PV and ProvisionerStatus Informer events.
+
+In the following description, we assume all Informer updates are cached and events buffered. On initialization, all caches are populated, and every Informer update event handler is executed once to ensure the entire provisioner system is in a consistent state.
+
+### ProvisionerStatus CRD
+The CRD contains the following fields (similar to design 1):
+
+* `LocalPVs` - A collection of PVs mounted in the local storage directory on the node. The discoverer is triggered to acquire this information.
+* `DeleteOperations` - A list keeping track of information related to each delete operation. This status is managed by the cleaner. It contains the following fields:
+  * PV - The PV to be deleted on disk.
+  * DeleteStatus - takes one of the following values:
+    * Requested
+    * InProgress
+    * Complete
+    * Error
+    
+There exists one ProvisionerStatus object per node.
```

Comment:
Will this approach address the security issue of a compromised Pod on a node modifying another node's object?  You could modify another node's DeleteStatus, which could cause the PV to get deleted and recreated prematurely.  You could also modify another node's LocalPVs, to return fake volumes or remove some existing volumes.

Is there a way we can restrict a single Pod in a DaemonSet to have permissions to only modify its object based on the Node its running on?

Reply:
@msau42 One possible way security can be maintained here is by having the provisioner also store the digital signature of some critical fields (DeleteOperations?) of the CRD in the CRD itself. The digital signature can be created using a secret key that is generated (one time activity) by the provisioner and stored locally in the node. The provisioner can "watch" its own CRD object. If it ever takes on a value that does not match the signature, then it can set itself into a "compromised" state (perhaps by writing a local file) where-in it will no longer cleans (deletes) volumes.  The only way this state can be reset is when an  admin manually deletes the file.  This admin intervention should never be needed unless there was a real security compromise or if there was a bug in provisioner code.

Reply:
Had a conversation with @liggitt offline. One possible solution is to give each pod its own service account, with write access only to its provisioner object's status. This wouldn't work with a DaemonSet and requires each pod run as something like a Deployment.

Alternatively we can have a hybrid approach with both CRD and gRPC: Only the master can write to CRDs, and minion provisioner communicates with master through gRPC to provide status updates.

If we use gRPC, we'll need to have a notion of container identity + authentication. Otherwise a malicious pod can claim to be another provisioner reporting false LocalVolumes, and there's no way for master to differentiate between a real provisioner and an impostor.

@dhirajh A malicious pod can still disrupt the entire local provisioner system, making local storage unusable. Admin can set an alert on provisioner objects becoming "compromised", but ideally we should make it impossible to have any disruption.

Reply:
> If we use gRPC, we'll need to have a notion of container identity + authentication. Otherwise a malicious pod can claim to be another provisioner reporting false LocalVolumes, and there's no way for master to differentiate between a real provisioner and an impostor.

only if you go with a push model. if the master pulls from worker pods, it can take status from the worker on node *n* and update the correct object

Reply:
The service account approach seemed interesting. I know that typically pods declare their service account in their manifest and then it gets used automatically in all their API requests. This is why I guess a daemonset cannot be used. Is there no way to allow a Pod to come up and then pick a service account after API access?

Reply:
If workers expose a Service in headless mode, then the Endpoints object is a reliable source of identification since it contains an IP address and its corresponding pod & node name.

It can be used for both push and pull models. For the push model, master extracts the source IP from the request and finds the matching node name in the Endpoints object.

Reply:
Do we need to assume IPs can be spoofed, or does Kubernetes already protect against that somehow?

Reply:
Regarding securing CRDs: if we make the resource non-namespaced instead, wouldn't it be difficult enough to meddle with the resource since most users/service accounts shouldn't be allowed to access non-namespaced resources anyway?